### PR TITLE
Makes mechs more vulnerable to some antagonists, chatgpt edition

### PR DESCRIPTION
**Pull request by githubuser4141**
## About The Pull Request

This is supposed to be an altnerative to simply nerfing the max armor of mechs - they retain their regular tankiness, but some antagonists have better effect on them, hopefully this is a more "holistic" approach, instead of making mechs be "weak against everything", make them be weaker against the strong opponents that should logically be a counter to them. 

Supposed to be an alternative to:
#88789 
and
#88781 (by allowing some blobs to specialize in anti-mech, rather than all mobs having massively increased damage vs mechs. Also increases the damage done by blob nauts.)

Also, removes some "cheat codes" that mechs have added to them years ago that people seem to've forgotten about or overlooked that makes them more powerful than expected.

Note: Code isn't fully mine, most of the attacks are written by chatgpt / text generators, as far as I've seen it works in testing with some minor changes.

Balance with obj_damage might not be 100%, some mobs have very high obj_damage values but unimpressive melee upper/lower damage values. I've tried to get the text generator to add "diminishing returns" to obj_damage so some random mob with like, 100 obj damage and 10 mob damage doesn't oneshot the mech. Imho some of it's weird and should be better balanced.

WIP for now, still testing some of the changes, takes a while at the moment.

Pic 1 - prying a moving mech (it doesn't work)

image

Pic 2 - prying a stationary mech

image

## Why It's Good For The Game

- Nerfs the durand's sleep punch attack, it just knocks people over now, instead of rendering them unconscious. This should allow for better counterplay by regular traitors, even if one is heavily armored.
- The durand's shield is now disabled by EMP attacks, and can't be turned back on until the EMP stun is over, should prevent a durand from tanking too many attacks through EMP by its shield.
- Buffs the amount of damage mechs receive from mobs with obj_damage. Tarantula and breacher spiders should now be noticeably stronger against mechs, especially lightly-armored mechs and the sides or back of a mech.
- Breachers spiders and xenomorphs now both have a element that makes their hands act like a mech crowbar when not on combat intent. This should allow for better counterplay if spiders or xenomorphs manage to swarm a mech and keep it immobile, either by cramming it into a corner, or by triggering a servo malfunction.
- Reduces mech gear's HP from 300 to 75. This allows combatants to disable the mech's attacks by aiming for its arms. Previously, with a HP of 300, this was much more difficult to actually accomplish as it's 300 damage *after* armor. 

- EMP blobs now cause heavy EMPs on mechs, other blob attacks are moderately buffed.
- Mech punches no longer do 3x their actual damage to blobs. (Durand's 40-damage punch used to do 120 damage to a blob tile, now it does only 40. Since blob tiles have a 0.25x brute mult, that's going from a one-hit kill on regular 25 HP block, to requiring three punches.

- The Odysseus has its HP raised to 200 (from 120), an extra utility slot (2>3) and a special ability to crowbar flocks and unpowered doors.

## Changelog
:cl:
add: Mech prying "element" is added

balance: Durand's Shield is disabled by EMP

balance: Removes heavy mecha punches making people unconscious, replaces it with knockdown

balance: EMP Blobs now cause a heavy EMP on mechas, with a cooldown of 5 seconds, other blob attacks are moderately buffed. Mecha punches no longer have a 3x damage multiplier on blob tiles, this takes a one-hit from a Durand to a three-hit to destroy a basic blob tile.

balance: Reduces mech equipment's HP from 300 to 75, to allow for combatants to disable mechs' arms more easily.

balance: Mechas now take moderately more damage from mobs with high amounts of obj_damage.

/:cl:

### CODE REVIEW

**Comment by carlarctg** [comment]
ok i was ready to laugh at this but the suggested changes are pretty great

File: code/modules/mob/living/basic/space_fauna/spider/giant_spider/giant_spiders.dm [hunk 1]
```diff
@@ -343,6 +343,7 @@
 
 	AddElement(/datum/element/wall_tearer)
 	AddElement(/datum/element/web_walker, /datum/movespeed_modifier/below_average_web)
+	AddElement(/datum/element/mech_prying)
```
Comment (carlarctg) [inline code/modules/mob/living/basic/space_fauna/spider/giant_spider/giant_spiders.dm:346]: you need to add mech_prying.dm to the dme

Comment (githubuser4141) [reply code/modules/mob/living/basic/space_fauna/spider/giant_spider/giant_spiders.dm:346]: Oh pffft. Whoops

File: code/modules/vehicles/mecha/mecha_defense.dm [hunk 2]
```diff
@@ -104,8 +115,28 @@
 		log_combat(user, src, "punched", "hulk powers")
 
 /obj/vehicle/sealed/mecha/blob_act(obj/structure/blob/B)
-	log_message("Attack by blob. Attacker - [B].", LOG_MECHA, color="red")
-	take_damage(30, BRUTE, MELEE, 0, get_dir(src, B))
+	if (!B?.overmind?.blobstrain)
```
Comment (carlarctg) [inline code/modules/vehicles/mecha/mecha_defense.dm:118]: just do isnull?

File: code/modules/vehicles/mecha/mecha_defense.dm [hunk 3]
```diff
@@ -104,8 +115,28 @@
...
+		return take_damage(30, BRUTE, MELEE, 0, get_dir(src, B))
+
+	var/datum/blobstrain/strain = B.overmind.blobstrain
+	log_message("Attack by blob strain: [strain.name]. Attacker - [B].", LOG_MECHA, color="red")
+
+	var/damage = 50
+
+	switch (strain.type)
```
Comment (carlarctg) [inline code/modules/vehicles/mecha/mecha_defense.dm:126]: no need for a switch on only two checks. just looks weird (also the spaces are weird but thats my personal take)

**Comment by carlarctg** [comment]
would suggest a needs picture label

File: code/modules/vehicles/mecha/mecha_defense.dm (hunk 3 above)
Comment (githubuser4141) [reply code/modules/vehicles/mecha/mecha_defense.dm:126]: tysm, I just copied it from the generator haha. I was trying to add more strain interactions like acid on blazing oil and slime but after messing about for a bit I realized adding acid to mechs would be a bit more work

**Comment by timothymtorres** [comment]
I'm curious what your LLM prompt looked like? I'm assuming you used one of the models that scans a codebase for code suggestions.

File: code/modules/vehicles/mecha/mech_melee_attack.dm [hunk 4]
```diff
@@ -52,8 +52,10 @@
 	return take_damage(mecha_attacker.force * 3, mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker)) // multiplied by 3 so we can hit objs hard but not be overpowered against mobs.
 
 /obj/machinery/mech_melee_attack(obj/vehicle/sealed/mecha/mecha_attacker, mob/living/user)
-	if(!user.combat_mode)
-		return
+	if(!user.combat_mode && mecha_attacker.can_manipulate_doors && (istype(src, /obj/machinery/door) || istype(src, /obj/machinery/door/firedoor)))
+		var/obj/machinery/door/target_door = src
+		target_door.try_to_crowbar(null, user)
+		return TRUE
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:58]: Rather than casting src, you should implement `/obj/machinery/door/mech_melee_attack`

**Comment by zxaber** [comment]
Spiders and xenos are already strong against humans. Why is it bad that mechs give them trouble?

**Comment by githubuser4141** [comment]
> Spiders and xenos are already strong against humans. Why is it bad that mechs give them trouble?
> 
> Durand shield being disabled by EMP and the punch changes are fine though.

It's just letting xenos open the hatch if the mech is completely stationary, it's not a big buff at all.

**Comment by GuillaumePrata** [comment]
> The durand's shield is now disabled by EMP attacks, and can't be turned back on until the EMP stun is over, should prevent a durand from tanking too many attacks through EMP by its shield.

I have said it in other places, but we should take the other approach instead, remove Durand's natural armor and make their shield be the core of the mech's survivability and preferably a cooldown spell instead of a power drain.

EMPs already are the main weapon against mechs, adding more and more on top is just silly.
We need new counterplays to make mech combat interesting, like the example above, playing around Durand's shield uptime, instead of just yolo EMPs for everything.

**Comment by githubuser4141** [comment]
> I'm curious what your LLM prompt looked like? I'm assuming you used one of the models that scans a codebase for code suggestions.

Hm, nope. I use chatgpt and Claude directly, by pasting snippets of relevant code into the box and telling it to code something. I'm aware that it's super tunnel visioned like this and has literally no (minimal?) awareness of anything outside the code I paste in, lol.

**Comment by Vect0r2** [comment]
iirc there was an older PR that was coded by bing chat that there was a discussion on the ownership of the code- not sure how it ended.

**Comment by Vect0r2** [comment]
found it https://github.com/tgstation/tgstation/pull/74039

**Comment by LemonInTheDark** [comment]
I do not accept AI generated content. This project has existed for over a decade, and ideally will continue to exist for longer then a few more months. In service of that, it's important that at least one party understands how a pr works, otherwise you are at best just pushing the work onto our maintainers, or at worst pouring void into our repo. Putting aside the issues of authorship in a database as small as ours (this is almost for sure trained on either OUR code or code from a bay fork)

I warned you bout this but IDK if you noticed, but this is me saying that in full.

If you'd like to contribute I encourage you to learn DM and use our community as a resource, that is what it exists for.

**Comment by timothymtorres** [comment]
> I do not accept AI generated content.

I'm going to chime in here and point out that there are new LLMs in development that are being trained exclusively using CC0 content. (including code) They are expected to release sometime in 2025.

**Comment by LemonInTheDark** [comment]
> I'm going to chime in here and point out that there are new LLMs in development that are being trained exclusively using CC0 content. (including code) They are expected to release sometime in 2025.

That's neat but this does not address my issues with them.

**Comment by DrAmazing343** [comment]
Any “AI written” code pushes the work from the pull request “author” to the maintainer team to figure out what’s what where and god forbid there’s not dogshit code that the maintainers then have to fix themselves because, despite the best of intentions, the “author” is indeed pushing robot code onto real people.

**Comment by LSommers** [comment]
> I do not accept AI generated content. 

CITRUS MAN B-
...
...
based?

**Pull request closed without merging**